### DISCUSSION TITLE
I've got my problem running in TFC, but my loss function is still larger than I want. What should I do?

### DISCUSSION THREAD

**Discussion by leakec** [leakec/tfc#29] [Q&A] [answered]
This is one of the most frequent questions I get. Hence, I'm posting the response with some common things to try.

**Comment by leakec** [comment, marked as answer]
## Common errors
The first thing that may be keeping your loss function from minimizing sufficiently is human error.

1. Check your constrained expression. If this is incorrect, then you are not solving the problem you think you are, which can lead to large minimums in the loss function. The most common error I see now that we have automatic differentiation is that people make mistakes when deriving their constrained expression. The most foolproof way to check is to check your constraints in the same Python script you're running the code in. For example, if you had a constraint like y(3) = y(5), actually check that 3 and 5 produce the same result when used as inputs to the function y. If you find your constrained expression is not correct, I recommend giving our constrained expression solver a go: it is a symbolic solver that utilizes SymPy and comes with the TFC package. You can find a tutorial on it [here](https://tfc-documentation.readthedocs.io/en/latest/notebooks/CeSolver.html).
2. Once you are sure your constrained expression is correct, the next place to check is when deriving your loss function. Double-check each step and make sure the res function is what you want it to be.

## Choosing basis functions
Once you have checked the above are positive you're solving the correct problem, the next thing to address is the basis functions you are using to try and minimize the loss function. There are a few things you can try:

1. Increase the number of basis functions. You may need to also increase the number of collocation points to do this since the total number of basis functions (not the degree, but the total number of functions) should be lower than the total number of points to get a good generalized solution (you'll get overfitting otherwise). 
2. Try a different basis function set. In particular, if you've been using polynomial bases, I would recommend giving some of the X-TFC functions a shot, e.g., ELMTanh. In contrast, if you've only tried X-TFC basis functions, try using a polynomial basis instead. When using the X-TFC versions, make sure to use "lstsq" as your solver, since you are likely to get singularities in the matrix when using a large number of X-TFC basis functions. 

## Issues fitting the solution
Even with a good basis set and the correct problem, you still may have issues getting the solution you want. Here are some common reasons why.

1. Does your problem have a lot of local minima? In that case, try seeding the problem closer to the solution you want (if you have some knowledge about the problem). If not, you may need to explore other solvers, since least-squares solvers will head towards a local minimum.
2. Does your problem have sharp gradients? If this is the case, then smooth basis functions will not do a good job. Again, the X-TFC functions may help a bit here, but there are likely other tricks you'll need to play like splitting the domain. This is a more complex topic, so feel free to dig through the dissertations (links are provided in their respective folders under `examples/`.)
